aten/src/ATen/core/ivalue.cpp

```diff
@@ -132,6 +153,101 @@ bool IValue::overlaps(const IValue& rhs) const {
   return false;
 }
 
+bool operator!=(const IValue& lhs, const IValue& rhs) {
+  return !(lhs == rhs);
+}
+
+bool operator==(const IValue& lhs, const IValue& rhs) {
+  IValue eq = lhs.equals(rhs);
+  if (eq.isBool()) {
+    return eq.toBool();
+  }
+  // The only case we don't return bool is for tensor comparison. In Python,
+  // `bool()` is called on the return value of `__eq__` if the return value is
+  // not a boolean. Mimic that behavior here.
+  TORCH_INTERNAL_ASSERT(eq.isTensor());
+  return eq.toTensor().is_nonzero();
+}
+
+IValue IValue::equals(const IValue& rhs) const {
+  const IValue& lhs = *this;
+  switch (lhs.tag) {
+    case Tag::None:
+      // In Python you're not supposed to do this comparison apparently. Not
+      // sure if we should warn here or what
+      return rhs.isNone();
+    case Tag::Tensor:
+      if (!rhs.isTensor()){return false;}
+      return lhs.toTensor().eq(rhs.toTensor());
+    case Tag::Double:
+      return rhs.isDouble() && lhs.toDouble() == rhs.toDouble();
+    case Tag::Int:
+      return rhs.isInt() && lhs.toInt() == rhs.toInt();
+    case Tag::Bool:
+      return rhs.isBool() && lhs.toBool() == rhs.toBool();
+    case Tag::String:
+      return rhs.isString() && lhs.toStringRef() == rhs.toStringRef();
+    case Tag::GenericDict:
+      return rhs.isGenericDict() && lhs.toGenericDict() == rhs.toGenericDict();
+    case Tag::Tuple:
+      return rhs.isTuple() && lhs.toTuple() == rhs.toTuple();
+    case Tag::Device:
+      return rhs.isDevice() && lhs.toDevice() == rhs.toDevice();
+    case Tag::GenericList:
+      return rhs.isList() && lhs.toList() == rhs.toList();
+    case Tag::Blob:
+      return rhs.isBlob() && lhs.isAliasOf(rhs);
+    case Tag::Future:
+      return rhs.isFuture() && lhs.isAliasOf(rhs);
+    case Tag::RRef:
+      return rhs.isRRef() && lhs.isAliasOf(rhs);
+    case Tag::Object:
+      return rhs.isObject() && lhs.isAliasOf(rhs);
+    case Tag::PyObject:
+      return rhs.isPyObject() && lhs.isAliasOf(rhs);
+    case Tag::Capsule:
+      return rhs.isCapsule() && lhs.isAliasOf(rhs);
+    case Tag::Uninitialized:
+      TORCH_INTERNAL_ASSERT(
```

Comment:
But arguably the `equals` function might be used in internal contexts as well. It's not clear this should throw.